peak/family.py

```diff
@@ -187,6 +188,29 @@ def prev(self) -> T:
 
         return _REG_CACHE.setdefault(key, Register)
 
+    def gen_attr_register(self, T, init):
+        key = (type(self), T, init)
```

Comment:
Aren't some of the hwtype ADTs  un-cachable? Can you add a test for a register as an ADT or provide assertion checking for invalid types?

Reply:
ADTs are hashable